server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -114,157 +110,232 @@ protected boolean resolveIndex() {
+                }
+                if (result.getResultType() == Engine.Result.Type.MAPPING_UPDATE_REQUIRED) {
+                    mappingUpdater.updateMappings(result.getRequiredMappingUpdate(), primary.shardId(),
+                        context.getRequestToExecute().type(),
+                        new ActionListener<Void>() {
+                            @Override
+                            public void onResponse(Void v) {
+                                context.markAsRequiringMappingUpdate();
+                                waitForMappingUpdate.accept(
+                                    ActionListener.runAfter(new ActionListener<Void>() {
+                                        @Override
+                                        public void onResponse(Void v) {
+                                            assert context.requiresWaitingForMappingUpdate();
+                                            context.resetForExecutionForRetry();
+                                        }
+
+                                        @Override
+                                        public void onFailure(Exception e) {
+                                            context.failOnMappingUpdate(e);
+                                        }
+                                    }, () -> itemDoneListener.onResponse(null))
+                                );
+                            }
+
+                            @Override
+                            public void onFailure(Exception e) {
+                                onComplete(exceptionToResult(e, primary, isDelete, version), context, updateResult);
+                                // Requesting mapping update failed, so we don't have to wait for a cluster state update
+                                assert context.isInitial();
+                                itemDoneListener.onResponse(null);
+                            }
+                        });
+                    return false;
+                } else {
+                    onComplete(result, context, updateResult);
+                }
             } catch (Exception e) {
-                context.failOnMappingUpdate(e);
-            }
-            return;
-        }
-
-        assert context.isOperationExecuted();
-
-        if (opType == DocWriteRequest.OpType.UPDATE &&
-            context.getExecutionResult().isFailed() &&
-            isConflictException(context.getExecutionResult().getFailure().getCause())) {
-            final UpdateRequest updateRequest = (UpdateRequest) context.getCurrent();
-            if (context.getRetryCounter() < updateRequest.retryOnConflict()) {
-                context.resetForExecutionForRetry();
-                return;
+                onComplete(exceptionToResult(e, primary, isDelete, version), context, updateResult);
             }
+        } catch (Exception e) {
+            itemDoneListener.onFailure(e);
```

Comment:
@ywelsch I have a question here: shouldn't we handle this case (general exception in the whole body of this method, in the real world probably only https://github.com/elastic/elasticsearch/pull/39793/files#diff-720a796f6beda1dfa6af60b45ffe1010R238) the same way we handle the rejection when coming back from a mapping update and fail all subsequent unprocessed items?